nanofi/include/api/nanofi.h

```diff
@@ -68,60 +94,173 @@ typedef int c2_start_callback(char *);
+ */
+void free_standalone_processor(standalone_processor* processor);
 
 /**
-* Register your callback to received flow files that the flow failed to process
-* The flow file ownership is transferred to the caller!
-* The first callback should be registered before the flow is used. Can be changed later during runtime.
-*/
+ * Register your callback to received flow files that the flow failed to process
+ * The flow file ownership is transferred to the caller!
+ * The first callback should be registered before the flow is used. Can be changed later during runtime.
+ * @param flow flow the callback belongs to
+ * @param onerror_callback callback to execute in case of failure
+ * @return 0 in case of success, -1 otherwise (flow is already in use)
+ **/
 int add_failure_callback(flow *flow, void (*onerror_callback)(flow_file_record*));
 
-
 /**
-* Set failure strategy. Please use the enum defined in cstructs.h
-* Return values: 0 (success), -1 (strategy cannot be set - no failure callback added?)
-* Can be changed runtime.
-* The defailt strategy is AS IS.
-*/
+ * Set failure strategy. Please use the enum defined in cstructs.h
+ * Can be changed runtime.
+ * The default strategy is AS IS.
+ * @param flow the flow to set strategy for
+ * @param strategy the strategy to be set
+ * @return 0 (success), -1 (strategy cannot be set - no failure callback added?)
+ **/
 int set_failure_strategy(flow *flow, FailureStrategy strategy);
 
-int set_property(processor *, const char *, const char *);
+/**
+ * Set property for a processor
+ * @param processor the processor the property is set for
+ * @param name name of the property
+ * @param value value of the property
+ * @return 0 in case of success, -1 otherwise (the processor doesn't support such property)
+ **/
+int set_property(processor * processor, const char * name, const char * value);
 
-int set_standalone_property(standalone_processor*, const char*, const char *);
+/**
+ * Set property for a standalone processor
+ * @param processor the processor the property is set for
+ * @param name name of the property
+ * @param value value of the property
+ * @return 0 in case of success, -1 otherwise (the processor doesn't support such property)
+ **/
+int set_standalone_property(standalone_processor * processor, const char * name, const char * value);
 
-int set_instance_property(nifi_instance *instance, const char*, const char *);
+/**
+ * Set property for an instance
+ * @param instance the instance the property is set for
+ * @param name name of the property
+ * @param value value of the property
+ * @return 0 in case of success, -1 otherwise. Always succeeds unless instance or name is nullptr/emtpy.
+ **/
+int set_instance_property(nifi_instance *instance, const char * name, const char * value);
 
-int free_flow(flow *);
+/**
+ * Get a property. Should be used in custom processor logic callbacks.
+ * @attention The returned value transfers ownership, it's the callers responsibility to free it!
```

Comment:
Am I interpreting this correctly in that the library calls malloc for a new character array and copies the memory as opposed to requiring the API user do that allocation?  This doesn't seem standard in the way that C does this.

Reply:
Unfortunately yes, you are correct.
The problem is "getProperty" returning string (not reference), which is a temporary object, so c_str() cannot be used as a return value. 
That's why there is an "attention" mark here, otherwise it would require an additional callback or change the internal API (change getProperty)

Reply:
MINIFICPP-693 is a follow on for this.